src/Client/ClientBase.cpp

```diff
@@ -2270,6 +2250,19 @@ void ClientBase::processParsedSingleQuery(const String & full_query, const Strin
                     "Processing async inserts with both inlined and external data (from stdin or infile) is not supported");
         }
 
+        String query_to_execute;
+        /// An INSERT query may have the data that follows query text.
+        /// Send part of the query without data, because data will be sent separately.
+        /// But for asynchronous inserts we don't extract data, because it's needed
+        /// to be done on server side in that case (for coalescing the data from multiple inserts on server side).
+        if (insert && !is_async_insert_with_inlined_data)
+        {
+            chassert(full_query.data() <= insert->data && insert->end <= full_query.data() + full_query.size() && "INSERT query points to full_query");
```

Comment:
@al13n321 I've updated my patch finally (if it will pass CI, we can close https://github.com/ClickHouse/ClickHouse/pull/78493), but it has extra complexity, due to now we have to calculate the `query_to_execute` inside this function, so we need that `ASTPtr` will point to the same memory region, so I've converted all callers and related API to use `std::string_view`, so PTAL